[Bug] Unkown Field in KeystoneProject

Why when trying to get the projects i am getting the following exception:
Error logs:
```
com.fasterxml.jackson.databind.JsonMappingException: Can not deserialize instance of java.lang.String out of START_OBJECT token
```
```
org.openstack4j.openstack.identity.v3.domain.KeystoneProject$Projects["projects"]->java.util.ArrayList[0]->org.openstack4j.openstack.identity.v3.domain.KeystoneProject["options"])
```

Tempurary solution could be:
Add "options" to Json Annotation in class org.openstack4j.openstack.identity.v3.domain.KeystoneProject
 * Origin:
 ```
@JsonIgnoreProperties(value = "options" , ignoreUnknown = true)
 ```
 * fix
 ```
@JsonIgnoreProperties(value = {"extra", "options"} , ignoreUnknown = true)
 ```

The full Exception
```
ClientResponseException{message=Cannot deserialize instance of `java.lang.String` out of START_OBJECT token
 at [Source: (org.apache.http.client.entity.LazyDecompressingInputStream); line: 1, column: 350] (through reference chain: org.openstack4j.openstack.identity.v3.domain.KeystoneProject$Projects["projects"]->java.util.ArrayList[0]->org.openstack4j.openstack.identity.v3.domain.KeystoneProject["options"]), status=0, status-code=CODE_UNKNOWN}
        at org.openstack4j.connectors.httpclient.HttpResponseImpl.readEntity(HttpResponseImpl.java:137)
        at org.openstack4j.core.transport.HttpEntityHandler.handle(HttpEntityHandler.java:66)
        at org.openstack4j.connectors.httpclient.HttpResponseImpl.getEntity(HttpResponseImpl.java:67)
        at org.openstack4j.openstack.internal.BaseOpenStackService$Invocation.execute(BaseOpenStackService.java:225)
        at org.openstack4j.openstack.internal.BaseOpenStackService$Invocation.execute(BaseOpenStackService.java:207)
        at org.openstack4j.openstack.identity.v3.internal.ProjectServiceImpl.list(ProjectServiceImpl.java:64)
        at com.example.cluster.OSClusterConnector.listProjects(OSClusterConnector.scala:305)
        at com.example.cluster.OSClusterConnector.$anonfun$projects$1(OSClusterConnector.scala:181)
        at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:658)
        at scala.util.Success.$anonfun$map$1(Try.scala:255)
        at scala.util.Success.map(Try.scala:213)
        at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
        at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
        at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
        at java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
        at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
        at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
        at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize instance of `java.lang.String` out of START_OBJECT token
 at [Source: (org.apache.http.client.entity.LazyDecompressingInputStream); line: 1, column: 350] (through reference chain: org.openstack4j.openstack.identity.v3.domain.KeystoneProject$Projects["projects"]->java.util.ArrayList[0]->org.openstack4j.openstack.identity.v3.domain.KeystoneProject["options"])
        at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:59)
        at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1442)
        at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1216)
        at com.fasterxml.jackson.databind.DeserializationContext.handleUnexpectedToken(DeserializationContext.java:1126)
        at com.fasterxml.jackson.databind.deser.std.StringDeserializer.deserialize(StringDeserializer.java:63)
        at com.fasterxml.jackson.databind.deser.std.StringDeserializer.deserialize(StringDeserializer.java:10)
        at com.fasterxml.jackson.databind.deser.SettableAnyProperty.deserialize(SettableAnyProperty.java:153)
        at com.fasterxml.jackson.databind.deser.SettableAnyProperty.deserializeAndSet(SettableAnyProperty.java:134)
        at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.handleUnknownVanilla(BeanDeserializerBase.java:1564)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:294)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:286)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:245)
        at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.deserialize(CollectionDeserializer.java:27)
        at com.fasterxml.jackson.databind.deser.impl.FieldProperty.deserializeAndSet(FieldProperty.java:138)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
        at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
        at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:1719)
        at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1228)
        at org.openstack4j.connectors.httpclient.HttpResponseImpl.readEntity(HttpResponseImpl.java:134)
        ... 19 more
```

--- a/core-test/src/main/java/org/openstack4j/api/identity/v3/KeystoneDomainServiceTests.java
+++ b/core-test/src/main/java/org/openstack4j/api/identity/v3/KeystoneDomainServiceTests.java
@@ -5,7 +5,12 @@ import static org.testng.Assert.assertEquals;
 import org.openstack4j.api.AbstractTest;
 import org.openstack4j.api.Builders;
 import org.openstack4j.model.identity.v3.Domain;
+import org.openstack4j.openstack.identity.v3.domain.KeystoneDomain;
+import org.openstack4j.openstack.identity.v3.domain.KeystoneRole;
 import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.util.Collections;
 
 /**
  * Tests the Identity/Keystone API version 3 DomainService
@@ -62,7 +67,13 @@ public class KeystoneDomainServiceTests extends AbstractTest {
         assertEquals(updatedDomain.getId(), DOMAIN_ID);
         assertEquals(updatedDomain.getName(), DOMAIN_NAME);
         assertEquals(updatedDomain.getDescription(), DOMAIN_DESCRIPTION_UPDATED);
-
     }
 
+    @Test
+    public void domain_with_options() throws IOException {
+        respondWith("/identity/v3/domains_with_options.json");
+        Domain domain = osv3().identity().domains().list().stream().filter(r -> "Default".equals(r.getName())).findFirst().get();
+        assertEquals(domain.getId(), "default");
+        assertEquals(domain.getOptions(), Collections.emptyMap());
+    }
 }

--- a/core-test/src/main/java/org/openstack4j/api/identity/v3/KeystoneProjectServiceTests.java
+++ b/core-test/src/main/java/org/openstack4j/api/identity/v3/KeystoneProjectServiceTests.java
@@ -3,7 +3,9 @@ package org.openstack4j.api.identity.v3;
 import static org.testng.Assert.assertEquals;
 import static org.testng.AssertJUnit.assertNull;
 
+import java.io.IOException;
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
 
 import org.openstack4j.api.AbstractTest;
@@ -30,7 +32,6 @@ public class KeystoneProjectServiceTests extends AbstractTest {
     private static final String PROJECT_EXTRA_KEY_2 = "extra_key2";
     private static final String PROJECT_EXTRA_VALUE_2 = "value2";
     private static final List<String> TAGS = Arrays.asList("one", "two", "three");
-    private String PROJECT_ID;
 
     @Override
     protected Service service() {
@@ -65,7 +66,7 @@ public class KeystoneProjectServiceTests extends AbstractTest {
         assertEquals(newProject.getExtra(PROJECT_EXTRA_KEY_1), PROJECT_EXTRA_VALUE_1);
         assertEquals(newProject.getTags(), TAGS);
 
-        PROJECT_ID = newProject.getId();
+        final String PROJECT_ID = newProject.getId();
 
         respondWith(JSON_PROJECTS_GET_BYID);
 
@@ -90,7 +91,15 @@ public class KeystoneProjectServiceTests extends AbstractTest {
         respondWith(JSON_PROJECTS_GET_BY_NAME_EMPTY);
         Project project = osv3().identity().projects().getByName(PROJECT_NAME, PROJECT_DOMAIN_ID);
         assertNull(project);
-    
     }
 
+    @Test
+    public void project_with_options() throws IOException {
+        respondWith("/identity/v3/projects_with_options.json");
+        List<? extends Project> list = osv3().identity().projects().list();
+        assertEquals(list.size(), 3);
+        Project project = list.stream().filter(p -> "demo".equals(p.getName())).findFirst().get();
+        assertEquals(project.getId(), "600905d353a84b20b644d2fe55a21e8a");
+        assertEquals(project.getOptions(), Collections.emptyMap());
+    }
 }

--- a/core-test/src/main/java/org/openstack4j/api/identity/v3/KeystoneRoleServiceTests.java
+++ b/core-test/src/main/java/org/openstack4j/api/identity/v3/KeystoneRoleServiceTests.java
@@ -4,12 +4,15 @@ import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertFalse;
 import static org.testng.Assert.assertTrue;
 
+import java.io.IOException;
+import java.util.Collections;
 import java.util.List;
 
 import org.openstack4j.api.AbstractTest;
 import org.openstack4j.model.common.ActionResponse;
 import org.openstack4j.model.identity.v3.Role;
 import org.openstack4j.model.identity.v3.RoleAssignment;
+import org.openstack4j.openstack.identity.v3.domain.KeystoneRole;
 import org.testng.annotations.Test;
 
 /**
@@ -292,4 +295,11 @@ public class KeystoneRoleServiceTests extends AbstractTest {
         return (value != null && value.length() != 0);
     }
 
+    @Test
+    public void role_with_options() throws IOException {
+        respondWith("/identity/v3/roles_with_options.json");
+        Role role = osv3().identity().roles().list().stream().filter(r -> "admin".equals(r.getName())).findFirst().get();
+        assertEquals(role.getId(), "f768cc1132a649e7840b6d14f21d0bd8");
+        assertEquals(role.getOptions(), Collections.emptyMap());
+    }
 }

--- a/core-test/src/main/resources/identity/v3/domains_with_options.json
+++ b/core-test/src/main/resources/identity/v3/domains_with_options.json
@@ -1,0 +1,20 @@
+{
+  "domains": [
+    {
+      "description": "The default domain",
+      "links": {
+        "self": "http://127.0.0.1:5000/v3/domains/default"
+      },
+      "tags": [],
+      "enabled": true,
+      "options": {},
+      "id": "default",
+      "name": "Default"
+    }
+  ],
+  "links": {
+    "self": "http://127.0.0.1:5000/v3/domains?limit=500",
+    "previous": null,
+    "next": null
+  }
+}

--- a/core-test/src/main/resources/identity/v3/projects_with_options.json
+++ b/core-test/src/main/resources/identity/v3/projects_with_options.json
@@ -1,0 +1,51 @@
+{
+  "links": {
+    "self": "http://127.0.0.1:5000/v3/projects",
+    "previous": null,
+    "next": null
+  },
+  "projects": [
+    {
+      "is_domain": false,
+      "description": "Tenant for the openstack services",
+      "links": {
+        "self": "http://127.0.0.1:5000/v3/projects/10b40033bbef48f89fe838fef62398f0"
+      },
+      "tags": [],
+      "enabled": true,
+      "id": "10b40033bbef48f89fe838fef62398f0",
+      "parent_id": "default",
+      "options": {},
+      "domain_id": "default",
+      "name": "services"
+    },
+    {
+      "is_domain": false,
+      "description": "default tenant",
+      "links": {
+        "self": "http://127.0.0.1:5000/v3/projects/600905d353a84b20b644d2fe55a21e8a"
+      },
+      "tags": [],
+      "enabled": true,
+      "id": "600905d353a84b20b644d2fe55a21e8a",
+      "parent_id": "default",
+      "options": {},
+      "domain_id": "default",
+      "name": "demo"
+    },
+    {
+      "is_domain": false,
+      "description": "admin tenant",
+      "links": {
+        "self": "http://127.0.0.1:5000/v3/projects/8519dba9f4594f0f87071c87784a8d2c"
+      },
+      "tags": [],
+      "enabled": true,
+      "id": "8519dba9f4594f0f87071c87784a8d2c",
+      "parent_id": "default",
+      "options": {},
+      "domain_id": "default",
+      "name": "admin"
+    }
+  ]
+}

--- a/core-test/src/main/resources/identity/v3/roles_with_options.json
+++ b/core-test/src/main/resources/identity/v3/roles_with_options.json
@@ -1,0 +1,69 @@
+{
+  "links": {
+    "self": "http://127.0.0.1:5000/v3/roles?limit=500",
+    "previous": null,
+    "next": null
+  },
+  "roles": [
+    {
+      "description": null,
+      "links": {
+        "self": "http://127.0.0.1:5000/v3/roles/08597a07c6d14d378cd0d6708c20e5fe"
+      },
+      "options": {},
+      "id": "08597a07c6d14d378cd0d6708c20e5fe",
+      "domain_id": null,
+      "name": "reader"
+    },
+    {
+      "description": null,
+      "links": {
+        "self": "http://127.0.0.1:5000/v3/roles/2149cfabf0f44c5bbad249b1a38a625e"
+      },
+      "options": {},
+      "id": "2149cfabf0f44c5bbad249b1a38a625e",
+      "domain_id": null,
+      "name": "member"
+    },
+    {
+      "description": null,
+      "links": {
+        "self": "http://127.0.0.1:5000/v3/roles/790bcad86d7346cd91a7516e7bec8959"
+      },
+      "options": {},
+      "id": "790bcad86d7346cd91a7516e7bec8959",
+      "domain_id": null,
+      "name": "ResellerAdmin"
+    },
+    {
+      "description": null,
+      "links": {
+        "self": "http://127.0.0.1:5000/v3/roles/832e442d00434cea84195f7df09bb1bb"
+      },
+      "options": {},
+      "id": "832e442d00434cea84195f7df09bb1bb",
+      "domain_id": null,
+      "name": "SwiftOperator"
+    },
+    {
+      "description": null,
+      "links": {
+        "self": "http://127.0.0.1:5000/v3/roles/e3fee47e88114d6f9a8e37916ab45fcf"
+      },
+      "options": {},
+      "id": "e3fee47e88114d6f9a8e37916ab45fcf",
+      "domain_id": null,
+      "name": "_member_"
+    },
+    {
+      "description": null,
+      "links": {
+        "self": "http://127.0.0.1:5000/v3/roles/f768cc1132a649e7840b6d14f21d0bd8"
+      },
+      "options": {},
+      "id": "f768cc1132a649e7840b6d14f21d0bd8",
+      "domain_id": null,
+      "name": "admin"
+    }
+  ]
+}

--- a/core/src/main/java/org/openstack4j/model/identity/v3/Domain.java
+++ b/core/src/main/java/org/openstack4j/model/identity/v3/Domain.java
@@ -31,6 +31,11 @@ public interface Domain extends ModelEntity, Buildable<DomainBuilder> {
     String getName();
 
     /**
+     * Domain options.
+     */
+    Map<String, String> getOptions();
+
+    /**
      * @return the Links of the domain
      */
     Map<String, String> getLinks();

--- a/core/src/main/java/org/openstack4j/model/identity/v3/Project.java
+++ b/core/src/main/java/org/openstack4j/model/identity/v3/Project.java
@@ -78,7 +78,12 @@ public interface Project extends ModelEntity, Buildable<ProjectBuilder> {
      * @return value for the given key
      */
     String getExtra(String key);
-    
+
+    /**
+     * Project options.
+     */
+    Map<String, String> getOptions();
+
     /**
      * 
      * @return list of tags

--- a/core/src/main/java/org/openstack4j/model/identity/v3/Role.java
+++ b/core/src/main/java/org/openstack4j/model/identity/v3/Role.java
@@ -26,6 +26,11 @@ public interface Role extends ModelEntity, Buildable<RoleBuilder> {
     String getName();
 
     /**
+     * Get Role options.
+     */
+    Map<String, String> getOptions();
+
+    /**
      * @return the links of the role
      */
     Map<String, String> getLinks();

--- a/core/src/main/java/org/openstack4j/model/identity/v3/builder/DomainBuilder.java
+++ b/core/src/main/java/org/openstack4j/model/identity/v3/builder/DomainBuilder.java
@@ -27,6 +27,8 @@ public interface DomainBuilder extends Builder<DomainBuilder, Domain> {
      */
     DomainBuilder name(String name);
 
+    DomainBuilder options(Map<String, String> options);
+
     /**
      * @see Domain#getLinks()
      */

--- a/core/src/main/java/org/openstack4j/model/identity/v3/builder/ProjectBuilder.java
+++ b/core/src/main/java/org/openstack4j/model/identity/v3/builder/ProjectBuilder.java
@@ -48,7 +48,9 @@ public interface ProjectBuilder extends Builder<ProjectBuilder, Project> {
 	 */
 	ProjectBuilder name(String name);
 
-	/**
+    ProjectBuilder options(Map<String, String> options);
+
+    /**
 	 * 
 	 * @see Project#getLinks()
 	 */

--- a/core/src/main/java/org/openstack4j/model/identity/v3/builder/RoleBuilder.java
+++ b/core/src/main/java/org/openstack4j/model/identity/v3/builder/RoleBuilder.java
@@ -22,6 +22,8 @@ public interface RoleBuilder extends Builder<RoleBuilder, Role> {
      */
     RoleBuilder name(String name);
 
+    RoleBuilder options(Map<String, String> options);
+
     /**
      * @see Role#getLinks()
      */

--- a/core/src/main/java/org/openstack4j/openstack/identity/v3/domain/KeystoneDomain.java
+++ b/core/src/main/java/org/openstack4j/openstack/identity/v3/domain/KeystoneDomain.java
@@ -3,6 +3,7 @@ package org.openstack4j.openstack.identity.v3.domain;
 import java.util.List;
 import java.util.Map;
 
+import com.google.common.collect.Maps;
 import org.openstack4j.model.identity.v3.Domain;
 import org.openstack4j.model.identity.v3.builder.DomainBuilder;
 import org.openstack4j.openstack.common.ListResult;
@@ -23,6 +24,7 @@ public class KeystoneDomain implements Domain {
     @JsonProperty
     private String name;
     private String description;
+    private Map<String, String> options = Maps.newHashMap();
     private Map<String, String> links;
     private boolean enabled;
 
@@ -60,6 +62,11 @@ public class KeystoneDomain implements Domain {
     @Override
     public String getName() {
         return name;
+    }
+
+    @Override
+    public Map<String, String> getOptions() {
+        return options;
     }
 
     /**
@@ -142,6 +149,12 @@ public class KeystoneDomain implements Domain {
         @Override
         public DomainBuilder name(String name) {
             model.name = name;
+            return this;
+        }
+
+        @Override
+        public DomainBuilder options(Map<String, String> options) {
+            model.options = options;
             return this;
         }
 

--- a/core/src/main/java/org/openstack4j/openstack/identity/v3/domain/KeystoneProject.java
+++ b/core/src/main/java/org/openstack4j/openstack/identity/v3/domain/KeystoneProject.java
@@ -25,7 +25,7 @@ import com.google.common.collect.Maps;
  * @see <a href="http://developer.openstack.org/api-ref-identity-v3.html#projects-v3">API reference</a>
  */
 @JsonRootName("project")
-/** If we don't explicitly set extra as an ignore property, it will methods with @JsonAnyGetter/Setter will not work **/
+/* If we don't explicitly set extra as an ignore property, it will methods with @JsonAnyGetter/Setter will not work */
 @JsonIgnoreProperties(value = "extra" , ignoreUnknown = true)
 public class KeystoneProject implements Project {
 
@@ -39,6 +39,7 @@ public class KeystoneProject implements Project {
     @JsonProperty("domain_id")
     private String domainId;
     private String description;
+    private Map<String, String> options = Maps.newHashMap();
     @JsonIgnore
     private Map<String, String> links;
     @JsonProperty("parent_id")
@@ -46,8 +47,10 @@ public class KeystoneProject implements Project {
     private String subtree;
     private String parents;
     private Boolean enabled = true;
-    private Map<String, String> extra = Maps.newHashMap();
     private List<String> tags = Lists.newArrayList();
+
+    /** Extra API properties served */
+    private Map<String, String> extra = Maps.newHashMap();
 
     /**
      * @return the Project builder
@@ -103,6 +106,11 @@ public class KeystoneProject implements Project {
         return name;
     }
 
+    @Override
+    public Map<String, String> getOptions() {
+        return options;
+    }
+
     /**
      * {@inheritDoc}
      */
@@ -156,6 +164,7 @@ public class KeystoneProject implements Project {
     /**
      * {@inheritDoc}
      */
+    @Override
     public String getExtra(String key) {
         return extra.get(key);
     }
@@ -298,6 +307,12 @@ public class KeystoneProject implements Project {
         @Override
         public ProjectBuilder name(String name) {
             model.name = name;
+            return this;
+        }
+
+        @Override
+        public ProjectBuilder options(Map<String, String> options) {
+            model.options = options;
             return this;
         }
 

--- a/core/src/main/java/org/openstack4j/openstack/identity/v3/domain/KeystoneRole.java
+++ b/core/src/main/java/org/openstack4j/openstack/identity/v3/domain/KeystoneRole.java
@@ -3,6 +3,7 @@ package org.openstack4j.openstack.identity.v3.domain;
 import java.util.List;
 import java.util.Map;
 
+import com.google.common.collect.Maps;
 import org.openstack4j.model.identity.v3.Role;
 import org.openstack4j.model.identity.v3.builder.RoleBuilder;
 import org.openstack4j.openstack.common.ListResult;
@@ -27,6 +28,7 @@ public class KeystoneRole implements Role {
     @JsonProperty("domain_id")
     private String domainId;
     private Map<String, String> links;
+    private Map<String, String> options = Maps.newHashMap();
 
     @Override
     public RoleBuilder toBuilder() {
@@ -59,6 +61,11 @@ public class KeystoneRole implements Role {
     @Override
     public String getDomainId() {
         return domainId;
+    }
+
+    @Override
+    public Map<String, String> getOptions() {
+        return options;
     }
 
     /**
@@ -121,11 +128,16 @@ public class KeystoneRole implements Role {
         }
 
         @Override
+        public RoleBuilder options(Map<String, String> options) {
+            model.options = options;
+            return this;
+        }
+
+        @Override
         public RoleBuilder links(Map<String, String> links) {
             model.links = links;
             return this;
         }
-
     }
 
     public static class Roles extends ListResult<KeystoneRole> {

